src/transactions/FeeBumpTransactionFrame.cpp

```diff
@@ -0,0 +1,425 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "transactions/FeeBumpTransactionFrame.h"
+#include "crypto/Hex.h"
+#include "crypto/SHA.h"
+#include "crypto/SignerKey.h"
+#include "ledger/LedgerTxn.h"
+#include "ledger/LedgerTxnEntry.h"
+#include "ledger/LedgerTxnHeader.h"
+#include "main/Application.h"
+#include "transactions/SignatureChecker.h"
+#include "transactions/SignatureUtils.h"
+#include "transactions/TransactionUtils.h"
+#include "xdrpp/marshal.h"
+
+#include <numeric>
+
+namespace stellar
+{
+
+static TransactionEnvelope
+fromFeeBumpInnerTx(TransactionEnvelope const& envelope)
+{
+    TransactionEnvelope e(ENVELOPE_TYPE_TX);
+    e.v1() = envelope.feeBump().tx.innerTx.v1();
+    return e;
+}
+
+static void
+updateResult(TransactionResult& outerRes, TransactionResult const& innerRes)
+{
+    if (innerRes.result.code() == txSUCCESS)
+    {
+        outerRes.result.results() = innerRes.result.results();
+    }
+    else
+    {
+        outerRes.result.code(txINNER_FAILED);
+        auto& inner = outerRes.result.innerResult();
+        inner.feeCharged = 0;
+        inner.result.code(innerRes.result.code());
+        if (inner.result.code() == txFAILED)
+        {
+            inner.result.results() = innerRes.result.results();
+        }
+    }
+}
+
+FeeBumpTransactionFrame::FeeBumpTransactionFrame(
+    Hash const& networkID, TransactionEnvelope const& envelope)
+    : mEnvelope(envelope)
+    , mInnerTx(std::make_shared<TransactionFrame>(
+          networkID, fromFeeBumpInnerTx(envelope), false))
+    , mNetworkID(networkID)
+{
+}
+
+bool
+FeeBumpTransactionFrame::apply(Application& app, AbstractLedgerTxn& ltx,
+                               TransactionMeta& meta)
+{
+    SignatureChecker signatureChecker{ltx.loadHeader().current().ledgerVersion,
+                                      getContentsHash(),
+                                      mEnvelope.feeBump().signatures};
+
+    LedgerTxn ltxTx(ltx);
+    // when applying, a failure during tx validation means that
+    // we'll skip trying to apply operations but we'll still
+    // process the sequence number if needed
+    auto cv = commonValid(signatureChecker, ltxTx, 0, true);
+
+    auto signaturesValid = cv >= (ValidationType::kInvalidPostAuth) &&
+                           processSignatures(signatureChecker, ltxTx);
+
+    auto& txChanges =
+        meta.v() == 1 ? meta.v1().txChanges : meta.v2().txChangesBefore;
+    txChanges = ltxTx.getChanges();
+    ltxTx.commit();
+
+    bool res = false;
+    if (signaturesValid && cv == ValidationType::kFullyValid)
```

Comment:
I'm actually quite not sure I understand what you're doing here:
as the purpose of `FeeBumpTransaction` is to get the inner transaction in, I don't see a reason to not always call `mInnerTx->apply` (and this is what https://github.com/stellar/stellar-protocol/blob/master/core/cap-0015.md#application-and-results-1 says too).

Reply:
I think the only purpose of the other validation checks is to remove the one time signer on the `feeSource` account - but this is a best effort from what I see if we do it at apply time (signers on the `feeSource` account may have changed).

A better solution is to deal with the one time signers on fee bump transactions when we charge the fee, that way this `apply` code is a pass through to the `innerTx->apply` and we properly isolate the side effects of `feeBump`

Reply:
This is related to some of the bugs that I mentioned yesterday. Too much copy-paste from the old implementation of CAP-0015. I'm going to do some general work on `FeeBumpTransactionFrame`, I wouldn't recommend continue reviewing that code until I update.